components/board/client.go

```diff
@@ -253,6 +265,125 @@ func (dic *digitalInterruptClient) RemoveCallback(c chan Tick) {
 	panic(errUnimplemented)
 }
 
+func (c *client) StreamTicks(ctx context.Context, interrupts []string, ch chan Tick, extra map[string]interface{}) error {
+	c.streamMu.Lock()
+	defer c.streamMu.Unlock()
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return err
+	}
+	c.extra = ext
+
+	// Start the stream
+	c.streamRunning = true
+	c.activeBackgroundWorkers.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	c.cancelBackgroundWorkers = cancel
+	// Create a background go routine that connects to the server's stream
+	utils.PanicCapturingGo(func() {
+		defer c.activeBackgroundWorkers.Done()
+		c.connectTickStream(ctx, interrupts, ch)
+	})
+	c.mu.RLock()
+	ready := c.streamReady
+	c.mu.RUnlock()
+
+	// wait until the stream is ready to return
+	for !ready {
+		c.mu.RLock()
+		ready = c.streamReady
+		c.mu.RUnlock()
+		// wait for 50 ms before checking again
+		if !utils.SelectContextOrWait(ctx, 50*time.Millisecond) {
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+func (c *client) connectTickStream(ctx context.Context, interrupts []string, ch chan Tick) {
+	defer func() {
+		c.streamMu.Lock()
+		defer c.streamMu.Unlock()
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.streamCancel = nil
+		c.streamRunning = false
+		c.streamReady = false
+	}()
+
+	streamCtx, cancel := context.WithCancel(ctx)
+	c.streamCancel = cancel
+
+	for {
+		c.mu.Lock()
+		c.streamReady = false
+		c.mu.Unlock()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		req := &pb.StreamTicksRequest{
+			Name:     c.info.name,
+			PinNames: interrupts,
+		}
+
+		stream, err := c.client.StreamTicks(streamCtx, req)
+		if err != nil {
+			c.logger.CError(ctx, err)
+			if utils.SelectContextOrWait(ctx, 3*time.Second) {
+				continue
+			} else {
+				return
+			}
+		}
+
+		// repeatly receive from the stream
+		for {
+			select {
+			case <-ctx.Done():
+				c.logger.CError(ctx, err)
+				return
+			default:
+			}
+
+			c.mu.Lock()
+			c.streamReady = true
+			c.mu.Unlock()
+			streamResp, err := stream.Recv()
+			if err != nil && streamResp == nil {
```

Comment:
I was going to have a question about what to do if `streamResp` was non-nil but we had an error anyway, but I see that case is handled below. I'd roll these together so the question doesn't even arise:
```go
if err != nil {
    if streamResp == nil {
        logger.Debug()
        streamCancel()
        return
    }
    // Otherwise, log the error but continue on
    c.logger.CEerror()
} else {
    ...
}
```